UnboundLocalError when attempting to connect as user requiring a password

### What action do you want to perform

I attempt to connect to a password-protected server via NoopExecutor.

### What are the results

This gives a confusing error (I deleted the venv name from the output):
```
self = <pytest_postgresql.factories.NoopExecutor object at 0x7fbe5d4df760>
    @property
    def version(self):
        """Get postgresql's version."""
        if self._version:
            return self._version
        try:
>           connection = psycopg2.connect(
                dbname='postgres',
                user=self.user,
                host=self.host,
                port=self.port,
                options=self.options
            )
env/lib/python3.8/site-packages/pytest_postgresql/factories.py:58:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
dsn = "dbname=postgres user=dbuser host=localhost port=5432 options=''", connection_factory = None, cursor_factory = None
kwargs = {'dbname': 'postgres', 'host': 'localhost', 'options': '', 'port': 5432, ...}, kwasync = {}
    def connect(dsn=None, connection_factory=None, cursor_factory=None, **kwargs):
        """
        Create a new database connection.

        The connection parameters can be specified as a string:

            conn = psycopg2.connect("dbname=test user=postgres password=secret")

        or using a set of keyword arguments:

            conn = psycopg2.connect(database="test", user="postgres", password="secret")

        Or as a mix of both. The basic connection parameters are:

        - *dbname*: the database name
        - *database*: the database name (only as keyword argument)
        - *user*: user name used to authenticate
        - *password*: password used to authenticate
        - *host*: database host address (defaults to UNIX socket if not provided)
        - *port*: connection port number (defaults to 5432 if not provided)

        Using the *connection_factory* parameter a different class or connections
        factory can be specified. It should be a callable object taking a dsn
        argument.

        Using the *cursor_factory* parameter, a new default cursor factory will be
        used by cursor().

        Using *async*=True an asynchronous connection will be created. *async_* is
        a valid alias (for Python versions where ``async`` is a keyword).

        Any other keyword parameter will be passed to the underlying client
        library: the list of supported parameters depends on the library version.

        """
        kwasync = {}
        if 'async' in kwargs:
            kwasync['async'] = kwargs.pop('async')
        if 'async_' in kwargs:
            kwasync['async_'] = kwargs.pop('async_')

        if dsn is None and not kwargs:
            raise TypeError('missing dsn and no parameters')

        dsn = _ext.make_dsn(dsn, **kwargs)
>       conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
E       psycopg2.OperationalError: fe_sendauth: no password supplied
env/lib/python3.8/site-packages/psycopg2/__init__.py:126: OperationalError
During handling of the above exception, another exception occurred:
request = <SubRequest 'postgresql_external' for <Function test_use_the_db>>
    @pytest.fixture
    def postgresql_factory(request):
        """
        Fixture factory for PostgreSQL.

        :param FixtureRequest request: fixture request object
        :rtype: psycopg2.connection
        :returns: postgresql client
        """
        config = get_config(request)
        if not psycopg2:
            raise ImportError(
                'No module named psycopg2. Please install either '
                'psycopg2 or psycopg2-binary package for CPython '
                'or psycopg2cffi for Pypy.'
            )
        proc_fixture = request.getfixturevalue(process_fixture_name)

        # _, config = try_import('psycopg2', request)
        pg_host = proc_fixture.host
        pg_port = proc_fixture.port
        pg_user = proc_fixture.user
        pg_options = proc_fixture.options
        pg_db = db_name or config['dbname']

        with DatabaseJanitor(
>               pg_user, pg_host, pg_port, pg_db, proc_fixture.version
        ):
env/lib/python3.8/site-packages/pytest_postgresql/factories.py:282:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <pytest_postgresql.factories.NoopExecutor object at 0x7fbe5d4df760>
    @property
    def version(self):
        """Get postgresql's version."""
        if self._version:
            return self._version
        try:
            connection = psycopg2.connect(
                dbname='postgres',
                user=self.user,
                host=self.host,
                port=self.port,
                options=self.options
            )
            version = str(connection.server_version)
            self._version = parse_version(
                '.'.join([
                    version[i: i+2] for i in range(0, len(version), 2)
                    if int(version[i: i+2])
                ])
            )
            return self._version
        finally:
>           connection.close()
E           UnboundLocalError: local variable 'connection' referenced before assignment
env/lib/python3.8/site-packages/pytest_postgresql/factories.py:74: UnboundLocalError
```

Since https://github.com/ClearcodeHQ/pytest-postgresql/blob/master/src/pytest_postgresql/factories.py#L58 is in try/finally and raises an exception but no connection is assigned to be closed.

### What are the expected results

`connection.close()` runs only if we have a connection.

`psycopg2.OperationalError: fe_sendauth: no password supplied` is the only exception raised.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+unreleased
+-------
+
+- [bugfix] NoopExecutor connects to read version by context manager to proerly handle cases
+  where it can't connect to the server.
+
 2.2.1
 -------
 

--- a/src/pytest_postgresql/factories.py
+++ b/src/pytest_postgresql/factories.py
@@ -54,11 +54,11 @@ class NoopExecutor:  # pylint: disable=too-few-public-methods
         """Get postgresql's version."""
         if not self._version:
             with psycopg2.connect(
-                dbname='postgres',
-                user=self.user,
-                host=self.host,
-                port=self.port,
-                options=self.options
+                    dbname='postgres',
+                    user=self.user,
+                    host=self.host,
+                    port=self.port,
+                    options=self.options
             ) as connection:
                 version = str(connection.server_version)
                 self._version = parse_version(

--- a/src/pytest_postgresql/factories.py
+++ b/src/pytest_postgresql/factories.py
@@ -54,14 +54,14 @@ class NoopExecutor:  # pylint: disable=too-few-public-methods
         """Get postgresql's version."""
         if self._version:
             return self._version
-        try:
-            connection = psycopg2.connect(
-                dbname='postgres',
-                user=self.user,
-                host=self.host,
-                port=self.port,
-                options=self.options
-            )
+
+        with psycopg2.connect(
+            dbname='postgres',
+            user=self.user,
+            host=self.host,
+            port=self.port,
+            options=self.options
+        ) as connection:
             version = str(connection.server_version)
             self._version = parse_version(
                 '.'.join([
@@ -69,9 +69,7 @@ class NoopExecutor:  # pylint: disable=too-few-public-methods
                     if int(version[i: i+2])
                 ])
             )
-            return self._version
-        finally:
-            connection.close()
+            return self._versio
 
 
 def get_config(request):

--- a/src/pytest_postgresql/factories.py
+++ b/src/pytest_postgresql/factories.py
@@ -52,24 +52,22 @@ class NoopExecutor:  # pylint: disable=too-few-public-methods
     @property
     def version(self):
         """Get postgresql's version."""
-        if self._version:
-            return self._version
-
-        with psycopg2.connect(
-            dbname='postgres',
-            user=self.user,
-            host=self.host,
-            port=self.port,
-            options=self.options
-        ) as connection:
-            version = str(connection.server_version)
-            self._version = parse_version(
-                '.'.join([
-                    version[i: i+2] for i in range(0, len(version), 2)
-                    if int(version[i: i+2])
-                ])
-            )
-            return self._versio
+        if not self._version:
+            with psycopg2.connect(
+                dbname='postgres',
+                user=self.user,
+                host=self.host,
+                port=self.port,
+                options=self.options
+            ) as connection:
+                version = str(connection.server_version)
+                self._version = parse_version(
+                    '.'.join([
+                        version[i: i+2] for i in range(0, len(version), 2)
+                        if int(version[i: i+2])
+                    ])
+                )
+        return self._version
 
 
 def get_config(request):

